pyhap/characteristic.py

```diff
@@ -91,61 +84,57 @@ def __init__(self, display_name, type_id, properties, value=None, broker=None):
-            else:
-                self.value = HAP_FORMAT.DEFAULT[properties["Format"]]
-        self.broker = broker
+        self.broker = None
         self.setter_callback = None
+        self.value = self._get_default_value()
 
     def __repr__(self):
         """Return the representation of the characteristic."""
-        return "<characteristic display_name='{}' value={} properties={}>" \
+        return '<characteristic display_name={} value={} properties={}>' \
             .format(self.display_name, self.value, self.properties)
 
-    def set_value(self, value, should_notify=True, should_callback=True):
+    def set_value(self, value, should_notify=True):
         """Set the given raw value. It is checked if it is a valid value.
 
         :param value: The value to assign as this Characteristic's value.
         :type value: Depends on properties["Format"]
 
-        :param should_notify: Whether a the change should be sent to subscribed clients.
-            The notification is called _after_ the setter callback. Notify will be
-            performed if and only if the broker is set, i.e. not None.
+        :param should_notify: Whether a the change should be sent to
+            subscribed clients. Notify will be performed if the broker is set.
         :type should_notify: bool
-
-        :param should_callback: Whether to invoke the callback, if such is set. This
-            is useful in cases where you and HAP clients can both update the value and
-            you don't want your callback called when you set the value, but want it
-            called when clients do. Defaults to True.
-        :type should_callback: bool
-
-        :raise ValueError: When the value being assigned is not one of the valid values
-            for this Characteristic.
         """
-        if self.properties.get('ValidValues') and \
-                value not in self.properties['ValidValues'].values():
-            raise ValueError
+        _LOGGER.debug('%s: Set value to %s', self.display_name, value)
+        value = self.to_valid_value(value)
         self.value = value
-        if self.setter_callback is not None and should_callback:
-            self.setter_callback(value)
-        if should_notify and self.broker is not None:
+        if should_notify and self.broker:
             self.notify()
 
+    def _get_default_value(self):
+        """Helper method. Return default value for format."""
+        if self.properties.get('ValidValues'):
+            return min(self.properties['ValidValues'].values())
+        else:
+            value = HAP_FORMAT.DEFAULT[self.properties['Format']] 
+            return self.to_valid_value(value)
+
+    def to_valid_value(self, value):
+        """Perform validation and conversion to valid value"""
+        if self.properties.get('ValidValues'):
+            if value not in self.properties['ValidValues'].values():
+                raise ValueError('%s is not a valid value'.format(value))
```

Comment:
@ikalchev I was thinking about this. What would you think about an error log statement instead of an ValueError? The log statement would be easier to debug.

Reply:
Why do you think so?

Reply:
With errors you get the whole traceback and it might seem unintentional. Whereas with error log message this would be handled much cleaner.

I started thinking in which cases you would use an `raise Error`. It is only if you don't know how to properly handle it in your program. So you push the responsibility to the end user.

As an implementation for us: I thought about `logger.error` and `return None`. Adding an `check if is not None` to `set_value`, otherwise abort.

Reply:
How about `logger.error`, `return value # as is`. iOS will show no response or something if it gets an invalid value ( I think)

Reply:
I think that breaks HomeKit. I tried it some time ago and it seemed to cause issues. I believe if the value doesn't get updated, that should be hint enough to think something is wrong here.

Reply:
Why would you not want an Exception and traceback? If the value is invalid then the caller should capture it, and either log it or ignore it (py-HAP is just providing an API after all, not then entire application. The demo server could be updated to capture and log the exception).

If the iOS client makes a call and sets an invalid value, then the exception should also be raised, caught by `client_set_value()` and the appropriate status response sent back to iOS.

Reply:
I don't think we should had over the responsibility for that.  Especially because we would need a try block for every `set_value` call.

For `client_set_value()`: By definition are all values stored in `self.value` are valid values for HomeKit. So HomeKit can't set an invalid value to `HAP-python`. There is just the case were it's an invalid value for the object / sensor you control, but I think this is for #75